### PR TITLE
feat: add cycle sort algorithm

### PR DESCRIPTION
**Pull request by dalaoqi**
#### Description of Change

Added the code for Cycle sort algorithm.

#### Checklist

- [x] Added description of change
- [x] Added file name matches [File name guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#New-File-Name-guidelines)
- [x] Added tests and example, test must pass
- [x] Added documentation so that the program is self-explanatory and educational - [Doxygen guidelines](https://www.doxygen.nl/manual/docblocks.html)
- [x] Relevant documentation/comments is changed or added
- [x] PR title follows semantic [commit guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#Commit-Guidelines)
- [x] Search previous suggestions before making a new one, as yours may be a duplicate.
- [x] I acknowledge that all my contributions will be made under the project's license.

Notes: 

<a href="https://gitpod.io/#https://github.com/TheAlgorithms/C-Plus-Plus/pull/1252"></a>

<a href="https://gitpod.io/#https://github.com/TheAlgorithms/C-Plus-Plus/pull/1252"></a>

### CODE REVIEW

File: sorting/cycle_sort.cpp [hunk 1]
```diff
@@ -0,0 +1,75 @@
+/**
+ * @file
+ * @brief Implementation of [Cycle sort](https://en.wikipedia.org/wiki/Cycle_sort) algorithm
+ *
+ *@details
+ *Cycle Sort is a sorting algorithm that works in \f$O(n^2)\f$ time in best cas and works in \f$O(n^2)\f$ in worst case.
+ *If a element is already at its correct  position, do nothing.
+ *If a element is not at its correct position, we then need to move it to its correct position by computing the correct positions.Therefore, we should make sure the duplicate elements.
+ *
+ *@author [TsungHan Ho](https://github.com/dalaoqi)
```
Comment (Panquesito7) [inline sorting/cycle_sort.cpp:10]:
```suggestion
 * @details
 * Cycle Sort is a sorting algorithm that works in \f$O(n^2)\f$ time in best cas and works in \f$O(n^2)\f$ in worst case.
 * If a element is already at its correct  position, do nothing.
 * If a element is not at its correct position, we then need to move it to its correct position by computing the correct positions.Therefore, we should make sure the duplicate elements.
 *
 *@author [TsungHan Ho](https://github.com/dalaoqi)
```

File: sorting/cycle_sort.cpp [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+ */
+
+#include <iostream>
+#include <vector>
+
+void swap(int *a, int *b) {
```
Comment (Panquesito7) [inline sorting/cycle_sort.cpp:16]: Missing function documentation.

File: sorting/cycle_sort.cpp [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+    int tmp = 0;
+    tmp = *a;
+    *a = *b;
+    *b = tmp;
+}
+
+void cycleSort(std::vector<int> *arr) {
```
Comment (Panquesito7) [inline sorting/cycle_sort.cpp:23]: Function not documented.

File: sorting/cycle_sort.cpp [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+            }
+            // duplicate  elements
+            while (item == (*arr)[pos]) pos += 1;
+            swap(&item, &(*arr)[pos]);
+        }
+    }
+}
+
+/* Test function */
+void test() {
```
Comment (Panquesito7) [inline sorting/cycle_sort.cpp:62]: All test functions must be **static**.
```suggestion
static void test() {
```

File: sorting/cycle_sort.cpp [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+                }
+            }
+            // duplicate  elements
+            while (item == (*arr)[pos]) pos += 1;
+            swap(&item, &(*arr)[pos]);
+        }
+    }
+}
+
+/* Test function */
```
Comment (Panquesito7) [inline sorting/cycle_sort.cpp:61]:
```suggestion
/**
 * @brief Test implementations
 * @returns void
 */
```

File: sorting/cycle_sort.cpp [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+void test() {
+    // [506, 48, 123, 79, 0, 362, 951, 500, 0] return [0, 0, 48, 79, 123, 362, 500, 506, 951]
+    std::vector<int> array = {506, 48, 123, 79, 0, 362, 951, 500, 0};
+    std::vector<int> *arr = &array;
+    cycleSort(arr);
+    std::cout << "Sorted : ";
+    for (int i : *arr) std::cout << i << " ";
+}
+
+/* Main function */
```
Comment (Panquesito7) [inline sorting/cycle_sort.cpp:71]:
```suggestion
/**
 * @brief Main function
 * @returns 0 on exit
 */
```

**Review by Panquesito7 — CHANGES_REQUESTED** [review]

File: sorting/cycle_sort.cpp [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+ */
+
+#include <iostream>
+#include <vector>
+
+/**
+ * @brief Interchange a and b
+ * @param a the first entity
+ * @param b the second entity
+ * @return void
```
Comment (Panquesito7) [inline sorting/cycle_sort.cpp:20]:
```suggestion
 * @returns void
```

File: sorting/cycle_sort.cpp [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+    int tmp = 0;
+    tmp = *a;
+    *a = *b;
+    *b = tmp;
+}
+
+/**
+ * @brief The main function implements cycleSort
+ * @param arr array to be sorted
+ * @return void
```
Comment (Panquesito7) [inline sorting/cycle_sort.cpp:32]:
```suggestion
 * @returns void
```

**Review by Panquesito7 — CHANGES_REQUESTED** [review]
Please enable [GitHub Actions](https://docs.github.com/en/free-pro-team@latest/actions/learn-github-actions) in your repository of this fork in this link: https://github.com/dalaoqi/C-Plus-Plus/actions

File: sorting/cycle_sort.cpp [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+ * If a element is already at its correct  position, do nothing.
+ * If a element is not at its correct position, we then need to move it to its correct position by computing the correct positions.Therefore, we should make sure the duplicate elements.
+ *
+ * @author [TsungHan Ho](https://github.com/dalaoqi)
+ */
+
+#include <iostream>
+#include <vector>
+
+/**
```
Comment (Panquesito7) [inline sorting/cycle_sort.cpp:16]:
```suggestion
/**
 * @namespace sorting
 * @brief Sorting algorithms
 */
namespace sorting {
/**
 * @namespace cycle_sort
 * @brief Functions for [Cycle sort](https://en.wikipedia.org/wiki/Cycle_sort) algorithm
 */
namespace cycle_sort {
/**
```

File: sorting/cycle_sort.cpp [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+                if ((*arr)[i] < item) {
+                    pos += 1;
+                }
+            }
+            // duplicate  elements
+            while (item == (*arr)[pos]) pos += 1;
+            swap(&item, &(*arr)[pos]);
+        }
+    }
+}
```
Comment (Panquesito7) [inline sorting/cycle_sort.cpp:70]:
```suggestion
}
}  // namespace cycle_sort
}  // namespace sorting

```

File: sorting/cycle_sort.cpp [hunk 11]
```diff
@@ -0,0 +1,92 @@
...
+
+/**
+ * @brief Main function
+ * @returns 0 on exit
+ */
+int main() {
+    test();  // execute the test
+    return 0;
+}
```
Comment (Panquesito7) [inline sorting/cycle_sort.cpp:92]:
```suggestion
}

```

File: sorting/cycle_sort.cpp [hunk 12]
```diff
@@ -0,0 +1,92 @@
...
+
+/**
+ * @brief Test implementations
+ * @returns void
+ */
+static void test() {
+    // [506, 48, 123, 79, 0, 362, 951, 500, 0] return [0, 0, 48, 79, 123, 362, 500, 506, 951]
+    std::vector<int> array = {506, 48, 123, 79, 0, 362, 951, 500, 0};
+    std::vector<int> *arr = &array;
+    cycleSort(arr);
```
Comment (Panquesito7) [inline sorting/cycle_sort.cpp:80]:
```suggestion
    sorting::cycle_sort::cycleSort(arr);
```

**Review by Panquesito7 — CHANGES_REQUESTED** [review]

File: sorting/cycle_sort.cpp [hunk 13]
```diff
@@ -0,0 +1,104 @@
...
+    }
+}
+} // namespace cycle_sort
+} // namespace sorting
+
+/**
+ * @brief Test implementations
+ * @returns void
+ */
+static void test() {
```
Comment (Panquesito7) [inline sorting/cycle_sort.cpp:88]: Test implementations use `assert` (requires the `cassert` library) to verify the code works as expected.

**Review by Panquesito7 — CHANGES_REQUESTED** [review]

**Comment by dalaoqi** [comment]
How do I solve this problem? Thx
image

**Comment by Panquesito7** [comment]
> How do I solve this problem?
> image

Should be fixed now.

File: sorting/cycle_sort.cpp [hunk 14]
```diff
@@ -0,0 +1,119 @@
...
+ */
+
+#include <algorithm>
+#include <cassert>
+#include <iostream>
+#include <vector>
+
+/**
+ * @namespace sorting
+ * @brief Sorting algorithm
```
Comment (Panquesito7) [inline sorting/cycle_sort.cpp:20]:
```suggestion
 * @brief Sorting algorithms
```

File: sorting/cycle_sort.cpp [hunk 15]
```diff
@@ -0,0 +1,119 @@
...
+ * If a element is already at its correct  position, do nothing.
+ * If a element is not at its correct position, we then need to move it to its correct position by computing the correct positions.Therefore, we should make sure the duplicate elements.
+ *
+ * @author [TsungHan Ho](https://github.com/dalaoqi)
+ */
+
+#include <algorithm>
+#include <cassert>
+#include <iostream>
+#include <vector>
```
Comment (Panquesito7) [inline sorting/cycle_sort.cpp:16]:
```suggestion
#include <algorithm> /// for std::is_sorted
#include <cassert>   /// for assert
#include <iostream> /// for io operations
#include <vector>   /// for std::vector
```

**Review by Panquesito7 — APPROVED** [review]
Amazing work, I like how you've improved the code so well.
Code and documentation looks good and very refined; LGTM. 😄 👍

**Comment by dalaoqi** [comment]
Thanks for the review! 😄

File: sorting/cycle_sort.cpp [hunk 16]
```diff
@@ -0,0 +1,119 @@
...
+
+/**
+ * @brief The main function implements cycleSort
+ * @tparam T type of array
+ * @param arr array to be sorted
+ * @returns void
+ */
+template <typename T>
+void cycleSort(std::vector<T> *arr) {
+    for (int cycle_start = 0; cycle_start <= (*arr).size() - 1; cycle_start++) {
```
Comment (kvedala) [inline sorting/cycle_sort.cpp:51]:
```suggestion
    for (size_t cycle_start = 0; cycle_start <= arr->size() - 1; cycle_start++) {
```

File: sorting/cycle_sort.cpp [hunk 17]
```diff
@@ -0,0 +1,119 @@
...
+        }
+
+        // item is already in correct position
+        if (pos == cycle_start) {
+            continue;
+        }
+
+        // duplicate  elements
+        while (item == (*arr)[pos]) pos += 1;
+        swap(&item, &(*arr)[pos]);
```
Comment (kvedala) [inline sorting/cycle_sort.cpp:70]:
```suggestion
        std::swap(item, (*arr)[pos]);
```

File: sorting/cycle_sort.cpp [hunk 18]
```diff
@@ -0,0 +1,119 @@
...
+            pos = cycle_start;
+            // Find position where we put the element
+            for (int i = cycle_start + 1; i < (*arr).size(); i++) {
+                if ((*arr)[i] < item) {
+                    pos += 1;
+                }
+            }
+            // duplicate  elements
+            while (item == (*arr)[pos]) pos += 1;
+            swap(&item, &(*arr)[pos]);
```
Comment (kvedala) [inline sorting/cycle_sort.cpp:83]:
```suggestion
            std::swap(item, (*arr)[pos]);
```

File: sorting/cycle_sort.cpp [hunk 19]
```diff
@@ -0,0 +1,119 @@
...
+
+        // duplicate  elements
+        while (item == (*arr)[pos]) pos += 1;
+        swap(&item, &(*arr)[pos]);
+
+        // Rest of the  elements
+        while (pos != cycle_start) {
+            pos = cycle_start;
+            // Find position where we put the element
+            for (int i = cycle_start + 1; i < (*arr).size(); i++) {
```
Comment (kvedala) [inline sorting/cycle_sort.cpp:76]:
```suggestion
            for (size_t i = cycle_start + 1; i < (*arr).size(); i++) {
```

File: sorting/cycle_sort.cpp [hunk 20]
```diff
@@ -0,0 +1,119 @@
...
+}
+
+/**
+ * @brief The main function implements cycleSort
+ * @tparam T type of array
+ * @param arr array to be sorted
+ * @returns void
+ */
+template <typename T>
+void cycleSort(std::vector<T> *arr) {
```
Comment (kvedala) [inline sorting/cycle_sort.cpp:50]: a better approach might be as below. This is only a suggestion and is up to you if you want to implement.
```suggestion
std::vector<T> cycleSort(const std::vector<T> &in_arr) {
    std::vector<T> arr(in_arr); // create a copy of the input array, process it and return a new array. Thus, the input array does not get modified and no need for pointers :)
```

**Review by kvedala — APPROVED** [review]
👍

**Pull request merged**